src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+    }
+  }
+
+  /// \brief Return a SparseBlockRowViewConst of block-row i of the matrix.
+  ///
+  /// If row i does not belong to the matrix, return an empty view.
+  ///
+  /// The returned object \c view implements the following interface:
+  /// <ul>
+  /// <li> \c view.length is the number of entries (i.e. blocks) 
+  ///      in the block row </li>
+  /// <li> \c view.local_row_in_block_row(K, i) returns a nonconst pointer
+  ///      to the values of the ith local row in the k-th block of the block-row </li>
+  /// <li> \c view.full_row_in_block_row(i) returns a nonconst pointer
+  ///      to the values of the ith local row of the block-row </li>
+  /// <li> \c view.local_block_value(K, i, j) returns a nonconst reference
+  ///      to the value in the ith local row and jth local col 
+  ///      of the k-th block of the block-row </li>
+  /// <li> \c view.block(K) returns an unmanaged 2D strided Kokkos::View
+  ///      of the values of the k-th block of the block-row </li>
+  /// </ul>
+  ///
+  /// Users should not rely on the return type of this method.  They
+  /// should instead assign to 'auto'.
+  ///
+  KOKKOS_INLINE_FUNCTION
+  SparseBlockRowViewConst<BlockCrsMatrix> block_row_Const (const ordinal_type i) const {
+
+    const size_type start = graph.row_map(i); // total num blocks prior to this block-row
+    const ordinal_type count = static_cast<ordinal_type> (graph.row_map(i+1) - start); // num blocks in this row
+
+    if (count == 0) {
+      return SparseBlockRowViewConst<BlockCrsMatrix> (NULL, NULL, 1, 0);
+    } else {
+      return SparseBlockRowViewConst<BlockCrsMatrix> (values, graph.entries, blockDim(), count, start);
+    }
+  }
+
+private:
+  ordinal_type numCols_;
+  ordinal_type blockDim_; // TODO Assuming square blocks for now - add blockRowDim, blockColDim
+};
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+// Assumes rows is the row_map pointer for the BlockCrsMatrix graph (i.e. cum sum of number of blocks)
+//         cols is the entries pointer for the BlockCrsMatrix graph (colidx for block-row blocks)
+//         annz is the total number of non-zeros in the CrsMatrix (blockDim*blockDim*numBlocks)
+template< typename ScalarType , typename OrdinalType, class Device, class MemoryTraits, typename SizeType >
+void
+BlockCrsMatrix<ScalarType , OrdinalType, Device, MemoryTraits, SizeType >::
+import (const std::string &label,
+        const OrdinalType nrows,
+        const OrdinalType ncols,
+        const size_type annz,
+        ScalarType* val,
+        OrdinalType* rows,
+        OrdinalType* cols,
+        const OrdinalType blockDim)
+{
+  std::string str = label;
+  values = values_type (str.append (".values"), annz);
+
+  numCols_ = ncols;
+  blockDim_ = blockDim;
+
+  // FIXME (09 Aug 2013) CrsArray only takes std::vector for now.
```

Comment:
Let's fix this.

Reply:
The use for the `import` method here and what is should play well with are not clear to me (this was copied along from the `KokkosSparse::CrsMatrix` and with functionality modified for the block structure of the graph). Can we discuss this more? I'd be happy to fix it :)

Reply:
I would say, get rid of it.  If nothing tests it, then it's broken :-) .

Reply:
I renamed the function to `ctor_impl` to prevent problems and confusion, will get the actual impl performance tested and cleaned up based on those results.

Reply:
@mhoemmen @srajama1 I have a commit ready to push to change the internals for this `ctor_impl` routine and wanted to comment on the changes here in case I'm overlooking something.

Replace the whole impl with the following:
```
  numCols_ = ncols;
  blockDim_ = blockDim;
  typename values_type::HostMirror unman_val( val, annz );
  typename row_map_type::HostMirror unman_rows( rows, nrows+1);
  typename index_type::HostMirror unman_cols( cols, ncols );

  // Create temporary Views for row_map and entries because the StaticCrsGraph ctor requires View inputs
  values_type tmp_row_map("tmp_row_map", nrows+1);
  values_type tmp_entries("tmp_entries", ncols);

  // deep_copy uses Kokkos::MDRangePolicy which will perform the copy much more efficiently than simply doing so serially
  Kokkos::deep_copy( val, unman_val );
  Kokkos::deep_copy( tmp_row_map, unman_rows );
  Kokkos::deep_copy( tmp_entries, unman_cols );

  // Construct graph using the temp entries and row_map Views
  graph = staticcrsgraph_type( tmp_entries, tmp_row_map );
```

This will be more efficient by comparing the previous impl to the code above - 
  // Previous ctor_impl does the following:
  // 1. Compute row counts from the `rows` pointer
  // 2. Create a `StaticCrsGraph` which basically reconstructs the existing `rows` pointer (stored in `HostMirror` of the `row_map` member) by serially iterating over the row counts in host space, followed by a `deep_copy` to the `row_map` member; also allocate the `entries` View
  // 3. Create temporary `HostMirror` views for `values` and `entries` for copying
  // 4. Serially copy on the host the `val` and `cols` pointers to `values` and `entries` (resp.)
  // 5. `deep_copy` results to corresponding member Views

  // Instead, reduce unnecessary copying and computation:
  // 1. Wrap the `val`, `rows`, `cols` in unmanaged `HostMirror` Views
  // 2. Create temporary Views (managed) for `entries` and `row_map` - these will be assigned to the `BlockCrsMatrix` `graph` member via its constructor (passed by reference)
  // 3. `deep_copy` the above unmanaged HostMirror Views to corresponding Views
  // 4. Assign `row_map` and `entries` to `graph` member; when `ctor_impl` completes the ref count of these managed Views will drop to 1 rather than 2. 

This removes the unnecessary computation of the row counts only to recompute what's stored in the already-existing rows pointer entries to store in the host row_map View, and removes the serial copies from the crs pointers to corresponding host views; replaced by unmanaged view construction to wrap each pointer and the existing `deep_copy` calls remain.

Reply:
Looks good to me